api,agent,server,engine-schema: scalability improvements

shwstppr opened the pull request:
### Description

Following changes and improvements have been added:

- Improvements in handling of PingRoutingCommand

    1. Added global config - `vm.sync.power.state.transitioning`, default value: true, to control syncing of power states for transitioning VMs. This can be set to false to prevent computation of transitioning state VMs.
    2. Improved VirtualMachinePowerStateSync to allow power state sync for host VMs in a batch
    3. Optimized scanning stalled VMs

- Added option to set worker threads for capacity calculation using config - `capacity.calculate.workers`

- Added caching for account/use role API access with expiration after write can be configured using config - `dynamic.apichecker.cache.period`. If set to zero then there will be no caching. Default is 0.

- Added caching for account/use role API access with expiration after write set to 60 seconds.

- Added caching for some recurring DB retrievals

    1. CapacityManager - listing service offerings - beneficial in host capacity calculation
    2. LibvirtServerDiscoverer existing host for the cluster - beneficial for host joins
    3. DownloadListener - hypervisors for zone - beneficial for host joins
    5. VirtualMachineManagerImpl - VMs in progress- beneficial for processing stalled VMs during PingRoutingCommands

- Optimized MS list retrieval for agent connect

- Optimize finding ready systemvm template for zone

- Database retrieval optimisations - fix and refactor for cases where only IDs or counts are used mainly for hosts and other infra entities. Also similar cases for VMs and other entities related to host concerning background tasks

- Changes in agent-agentmanager connection with NIO client-server classes

    1. Optimized the use of the executor service
    2. Refactore Agent class to better handle connections.
    3. Do SSL handshakes within worker threads
    5. Added global configs to control the behaviour depending on the infra. SSL handshake and initial processing of a new agent could be a bottleneck during agent connections. Configs - `agent.max.concurrent.new.connections` can be used to control number of new connections management server handles at a time. `agent.ssl.handshake.timeout` can be used to set number of seconds after which SSL handshake times out at MS end.
    6. On agent side backoff and sslhandshake timeout can be controlled by agent properties. `backoff.seconds` and `ssl.handshake.timeout` properties can be used.

- Improvements in StatsCollection - minimize DB retrievals.

- Improvements in DeploymentPlanner allow for the retrieval of only desired host fields and fewer retrievals.

- Improvements in hosts connection for a storage pool. Added config - `storage.pool.host.connect.workers` to control the number of worker threads that can be used to connect hosts to a storage pool. Worker thread approach is followed currently only for NFS and ScaleIO pools.

- Minor improvements in resource limit calculations wrt DB retrievals

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI
- [ ] test (unit or integration test code)

### Feature/Enhancement Scale or Bug Severity

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial


### Screenshots (if appropriate):


### How Has This Been Tested?

<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, and the tests you ran to -->

#### How did you try to break this feature and the system with this change?

<!-- see how your change affects other areas of the code, etc. -->


<!-- Please read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/main/CONTRIBUTING.md) document -->

GutoVeronezi commented:
Honestly, I don't like PRs with thousand of lines doing thousand of things. It is hard to review and test. I encourage you to separate it in several minor PRs that address each one of the changes you are proposing.

shwstppr commented:
@blueorangutan package

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 11422

shwstppr commented:
@blueorangutan package

blueorangutan commented:
@shwstppr a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 11441

shwstppr commented:
@blueorangutan package

blueorangutan commented:
@shwstppr a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 11486

shwstppr commented:
@blueorangutan test

blueorangutan commented:
@shwstppr a [SL] Trillian-Jenkins test job (ol8 mgmt + kvm-ol8) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-11738)</b>
Environment: kvm-ol8 (x2), Advanced Networking with Mgmt server ol8
Total time taken: 71834 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr9840-t11738-kvm-ol8.zip
Smoke tests completed. 134 look OK, 6 have errors, 1 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_host_tags | `Failure` | 10.40 | test_host_tags.py
test_03_secured_to_nonsecured_vm_migration | `Error` | 1286.75 | test_vm_life_cycle.py
test_03_secured_to_nonsecured_vm_migration | `Error` | 1286.76 | test_vm_life_cycle.py
ContextSuite context=TestSecuredVmMigration>:teardown | `Error` | 2253.98 | test_vm_life_cycle.py
ContextSuite context=TestVAppsVM>:setup | `Error` | 2409.13 | test_vm_life_cycle.py
ContextSuite context=TestVMLifeCycle>:setup | `Error` | 2499.28 | test_vm_life_cycle.py
ContextSuite context=TestVMSchedule>:setup | `Error` | 0.00 | test_vm_schedule.py
test_hostha_enable_ha_when_host_in_maintenance | `Error` | 2101.84 | test_hostha_kvm.py
test_01_migrate_vm_strict_tags_success | `Error` | 99.22 | test_vm_strict_host_tags.py
test_01_redundant_vpc_site2site_vpn | `Failure` | 455.29 | test_vpc_vpn.py
all_test_vm_lifecycle_unmanage_import | `Skipped` | --- | test_vm_lifecycle_unmanage_import.py

shwstppr commented:
@blueorangutan package

weizhouapache reviewed (approved):
great job @shwstppr 

over lgtm

it seems this will reduce a large number of database queries, therefore improve the performance a lot.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 11496

shwstppr commented on agent/src/main/java/com/cloud/agent/Agent.java:
```diff
@@ -1136,22 +1175,16 @@ public void doTask(final Task task) throws TaskExecutionException {
                     logger.error("Error parsing task", e);
                 }
             } else if (task.getType() == Task.Type.DISCONNECT) {
-                try {
```
Removed this because reconnect method has been refactored to allow waiting for backoff.seconds (default = 5 sec) before connecting to the next host

weizhouapache replied:
I see, thanks

shwstppr commented on engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:
```diff
@@ -223,17 +231,20 @@ public boolean configure(final String name, final Map<String, Object> params) th
 
         // executes all agent commands other than cron and ping
-        _directAgentExecutor = new ScheduledThreadPoolExecutor(DirectAgentPoolSize.value(), new NamedThreadFactory("DirectAgent"));
+        _directAgentExecutor = new ThreadPoolExecutor(Math.max(agentTaskThreads/10, 1), DirectAgentPoolSize.value(), 120L, TimeUnit.SECONDS, new LinkedBlockingQueue<>(), new NamedThreadFactory("DirectAgent"));
```
Used 1/10th because default value of config - DirectAgentPoolSize is 500 and it was creating that many core threads from

weizhouapache replied:
thanks for the checking
resolved

shwstppr commented on engine/schema/src/main/java/com/cloud/capacity/dao/CapacityDaoImpl.java:
```diff
@@ -671,6 +671,18 @@ public CapacityVO findByHostIdType(Long hostId, short capacityType) {
 
+    @Override
+    public List<CapacityVO> listByHostIdTypes(Long hostId, List<Short> capacityTypes) {
+        SearchBuilder<CapacityVO> sb = createSearchBuilder();
```
This searchbuilder is used only by this method, would it add any benefit in adding it as a class object?

weizhouapache replied:
not much benifit
we only need to initialize a final variable only once

shwstppr commented on engine/schema/src/main/java/com/cloud/dc/ClusterDetailsDaoImpl.java:
```diff
@@ -82,6 +85,23 @@ public Map<String, String> findDetails(long clusterId) {
+        if (CollectionUtils.isEmpty(names)) {
+            return new HashMap<>();
+        }
+        SearchBuilder<ClusterDetailsVO> sb = createSearchBuilder();
```
This searchbuilder is used only by this method, would it add any benefit in adding it as a class object?

shwstppr commented on engine/schema/src/main/java/com/cloud/dc/dao/ClusterDaoImpl.java:
```diff
@@ -167,23 +171,17 @@ public List<HypervisorType> getAvailableHypervisorInZone(Long zoneId) {
         }
         List<ClusterVO> clusters = listBy(sc);
-        List<HypervisorType> hypers = new ArrayList<HypervisorType>(4);
+        List<HypervisorType> distinctHypervisors = new ArrayList<>(4);
```
removed

shwstppr commented on engine/schema/src/main/java/com/cloud/host/dao/HostDaoImpl.java:
```diff
@@ -1602,4 +1701,75 @@ private String createSqlFindHostToExecuteCommand(boolean useDisabledHosts) {
+
+    @Override
+    public List<HypervisorType> listDistinctHypervisorTypes(final Long zoneId) {
+        GenericSearchBuilder<HostVO, HypervisorType> sb = createSearchBuilder(HypervisorType.class);
```
This searchbuilder is used only by this method, would it add any benefit in adding it as a class object?

shwstppr commented on engine/schema/src/main/java/org/apache/cloudstack/storage/datastore/db/PrimaryDataStoreDao.java:
```diff
@@ -154,4 +154,6 @@ Pair<List<Long>, Integer> searchForIdsAndCount(Long storagePoolId, String storag
 
     List<StoragePoolVO> listByIds(List<Long> ids);
+
+    List<Long> listAllIds();
```
Moved to GenericDaoBase (though code may not be too pretty)

weizhouapache replied:
thanks

shwstppr commented on plugins/hypervisors/simulator/src/main/java/com/cloud/resource/AgentRoutingResource.java:
```diff
@@ -122,7 +123,8 @@ public PingCommand getCurrentStatus(long id) {
             }
 
-            config = _simMgr.getMockConfigurationDao().findByNameBottomUP(agentHost.getDataCenterId(), agentHost.getPodId(), agentHost.getClusterId(), agentHost.getId(), "PingRoutingWithNwGroupsCommand");
+            //config = _simMgr.getMockConfigurationDao().findByNameBottomUP(agentHost.getDataCenterId(), agentHost.getPodId(), agentHost.getClusterId(), agentHost.getId(), "PingRoutingWithNwGroupsCommand");
```
removed

shwstppr commented on server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:
```diff
@@ -2428,7 +2428,7 @@ protected void checkIfZoneIsDeletable(final long zoneId) {
 
         // Check if there are any non-removed hosts in the zone.
-        if (!_hostDao.listByDataCenterId(zoneId).isEmpty()) {
+        if (!_hostDao.listEnabledIdsByDataCenterId(zoneId).isEmpty()) {
```
Original method listed only Enabled hosts,
https://github.com/apache/cloudstack/blob/main/engine/schema/src/main/java/com/cloud/host/dao/HostDaoImpl.java#L514-L517

Should we list all including disabled?

weizhouapache replied:
OK, if the current behavior is same,  let's keep it

shwstppr commented on engine/schema/src/main/java/com/cloud/storage/dao/StoragePoolHostDaoImpl.java:
```diff
@@ -169,23 +169,23 @@ public List<Long> findHostsConnectedToPools(List<Long> poolIds) {
             ResultSet rs = pstmt.executeQuery();
             while (rs.next()) {
-                l.add(new Pair<Long, Integer>(rs.getLong(1), rs.getInt(2)));
+                poolCount = rs.getLong(1);
```
Since the method is used to find if there is a connected storage pool and both SQLs are listing IDs of storage_pool_host_ref, this should be okay.
We can otherwise add those IDs to a list and check if the list is not empty but it would return the same result.

shwstppr commented on server/src/main/java/com/cloud/storage/StorageManagerImpl.java:
```diff
@@ -1540,6 +1545,77 @@ protected String getStoragePoolNonDestroyedVolumesLog(long storagePoolId) {
+
+    @Override
+    public void connectHostsToPool(DataStore primaryStore, List<Long> hostIds, Scope scope,
+              boolean handleExceptionsPartially, boolean errorOnNoUpHost) throws CloudRuntimeException {
```
Added check

shwstppr commented on server/src/main/java/org/apache/cloudstack/acl/RoleManagerImpl.java:
```diff
@@ -594,7 +594,7 @@ public String getConfigComponentName() {
     @Override
     public ConfigKey<?>[] getConfigKeys() {
-        return new ConfigKey<?>[] {RoleService.EnableDynamicApiChecker};
+        return new ConfigKey<?>[] {RoleService.EnableDynamicApiChecker, RoleService.DynamicApiCheckerCachePeriod};
```
done

shwstppr commented on tools/marvin/setup.py:
```diff
@@ -27,7 +27,7 @@
 
 
-VERSION = "4.20.0.0-SNAPSHOT"
+VERSION = "4.20.0.0"
```
done

shwstppr commented on utils/src/main/java/com/cloud/utils/nio/NioConnection.java:
```diff
@@ -190,9 +209,25 @@ public Boolean call() throws NioConnectionException {
     abstract void unregisterLink(InetSocketAddress saddr);
 
+    protected boolean rejectConnectionIfBusy(final SocketChannel socketChannel) throws IOException {
+        if (activeAcceptConnections.get() < sslHandshakeMaxWorkers) {
```
Wait and retry is present at the client side (agent).
(Although in a multi-managment server setup, the agent may end up connecting to a different MS in this case)

weizhouapache replied:
ok, that's good enough

shwstppr commented on agent/src/main/java/com/cloud/agent/Agent.java:
```diff
@@ -444,59 +466,87 @@ protected void cancelTasks() {
-        InetAddress addr;
+    protected String retrieveHostname() {
+        if (logger.isTraceEnabled()) {
+            logger.trace(" Retrieving hostname " + serverResource.getClass().getSimpleName());
```
done

shwstppr commented on plugins/hypervisors/simulator/src/main/java/com/cloud/resource/AgentRoutingResource.java:
```diff
@@ -109,7 +109,8 @@ public Type getType() {
         TransactionLegacy txn = TransactionLegacy.open(TransactionLegacy.SIMULATOR_DB);
         try {
-            MockConfigurationVO config = _simMgr.getMockConfigurationDao().findByNameBottomUP(agentHost.getDataCenterId(), agentHost.getPodId(), agentHost.getClusterId(), agentHost.getId(), "PingCommand");
+            //MockConfigurationVO config = _simMgr.getMockConfigurationDao().findByNameBottomUP(agentHost.getDataCenterId(), agentHost.getPodId(), agentHost.getClusterId(), agentHost.getId(), "PingCommand");
```
removed

shwstppr commented on engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:
```diff
@@ -4988,89 +5016,58 @@ private void scanStalledVMInTransitionStateOnDisconnectedHosts() {
-                "AND (i.state ='Starting' OR i.state='Stopping' OR i.state='Migrating') " +
-                "AND i.id NOT IN (SELECT w.vm_instance_id FROM vm_work_job AS w JOIN async_job AS j ON w.id = j.id WHERE j.job_status = ?)" +
-                "AND i.removed IS NULL";
+        final String sql = "SELECT i.*\n" +
```
done

shwstppr commented on framework/db/src/main/java/com/cloud/utils/db/TransactionLegacy.java:
```diff
@@ -1334,7 +1334,7 @@ private static DataSource createDbcpDataSource(String uri, String username, Stri
         }
         return new PoolingDataSource<>(connectionPool);
-    }
+     }
```
done

shwstppr commented on server/src/main/java/com/cloud/alert/AlertManagerImpl.java:
```diff
@@ -257,6 +259,64 @@ public void sendAlert(AlertType alertType, long dataCenterId, Long podId, String
+            return;
+        }
+        ConcurrentHashMap<Long, Future<Void>> futures = new ConcurrentHashMap<>();
+        ExecutorService executorService = Executors.newFixedThreadPool(Math.max(1,
```
It uses Math.min(value of `capacity.calculate.workers` and size of hosts/pools). Config can be set a negative value (though edge case and operator error) and when Executors.newFixedThreadPool is instantiated with <1 it may produce exception

weizhouapache replied:
ok, makes sense

shwstppr commented on api/src/main/java/org/apache/cloudstack/api/command/admin/domain/ListDomainsCmd.java:
```diff
@@ -142,7 +142,10 @@ protected void updateDomainResponse(List<DomainResponse> response) {
-        _resourceLimitService.updateTaggedResourceLimitsAndCountsForDomains(response, getTag());
+        EnumSet<DomainDetails> details = getDetails();
+        if (details.contains(DomainDetails.all) || details.contains(DomainDetails.resource)) {
+            _resourceLimitService.updateTaggedResourceLimitsAndCountsForDomains(response, getTag());
```
This PR is intended for 4.20.1. I can create a separate PR for these changes but not sure if we can get it in 4.20.0 release with RC/release work already underway

weizhouapache replied:
ok, 4.20.1 is good to go then

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 11505

shwstppr commented:
@blueorangutan test

blueorangutan commented:
@shwstppr a [SL] Trillian-Jenkins test job (ol8 mgmt + kvm-ol8) has been kicked to run smoke tests

weizhouapache commented on agent/src/main/java/com/cloud/agent/Agent.java:
```diff
@@ -444,59 +466,87 @@ protected void cancelTasks() {
+        if (logger.isTraceEnabled()) {
+            logger.trace(" Retrieving hostname " + serverResource.getClass().getSimpleName());
+        }
+        final String result = Script.runSimpleBashScript(Script.getExecutableAbsolutePath("hostname"), 500);
```
that should be fine, thanks

weizhouapache commented on api/src/main/java/org/apache/cloudstack/api/command/user/account/ListAccountsCmd.java:
```diff
@@ -149,7 +149,10 @@ protected void updateAccountResponse(List<AccountResponse> response) {
-        _resourceLimitService.updateTaggedResourceLimitsAndCountsForAccounts(response, getTag());
+        EnumSet<DomainDetails> details = getDetails();
+        if (details.contains(DomainDetails.all) || details.contains(DomainDetails.resource)) {
+            _resourceLimitService.updateTaggedResourceLimitsAndCountsForAccounts(response, getTag());
```
it is good if we get this in 4.20.1

weizhouapache commented on engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:
```diff
@@ -3793,12 +3803,12 @@ public int getTimeout() {
     public boolean processCommands(final long agentId, final long seq, final Command[] cmds) {
         boolean processed = false;
         for (final Command cmd : cmds) {
+            // FIXME: PingRoutingCommand handler is DB & CPU hotspot
```
thanks
good to know it

weizhouapache commented:
@blueorangutan test

blueorangutan commented:
@weizhouapache a [SL] Trillian-Jenkins test job (ol8 mgmt + kvm-ol8) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-11749)</b>
Environment: kvm-ol8 (x2), Advanced Networking with Mgmt server ol8
Total time taken: 64218 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr9840-t11749-kvm-ol8.zip
Smoke tests completed. 136 look OK, 5 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_02_list_cpvm_vm | `Failure` | 0.03 | test_ssvm.py
test_04_cpvm_internals | `Failure` | 0.03 | test_ssvm.py
test_01_migrate_VM_and_root_volume | `Error` | 100.30 | test_vm_life_cycle.py
test_02_migrate_VM_with_two_data_disks | `Error` | 54.63 | test_vm_life_cycle.py
test_08_migrate_vm | `Error` | 50.01 | test_vm_life_cycle.py
test_01_migrate_vm_strict_tags_success | `Error` | 105.19 | test_vm_strict_host_tags.py
test_01_verify_ipv6_vpc | `Error` | 1704.96 | test_vpc_ipv6.py
test_01_create_redundant_VPC_2tiers_4VMs_4IPs_4PF_ACL | `Failure` | 622.70 | test_vpc_redundant.py

blueorangutan commented:
<b>[SF] Trillian test result (tid-11754)</b>
Environment: kvm-ol8 (x2), Advanced Networking with Mgmt server ol8
Total time taken: 57015 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr9840-t11754-kvm-ol8.zip
Smoke tests completed. 140 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_sys_vm_start | `Failure` | 0.13 | test_secondary_storage.py

harikrishna-patnala reviewed (approved):
minor comment, overall LGTM. Nice work @shwstppr

shwstppr commented:
Thank you @harikrishna-patnala for the review

@GutoVeronezi thank you for the feedback and suggestion. I agree with the rationale of smaller PRs but I would request you and others to consider this as an exception. The work in this PR is a part of larger scalability work which I'll also be presenting at CCC. In this PR, the focus is mainly on persistence layer and infrastructure entities. There are also some changes wrt schema and other areas which were quick fixes and yield performance improvements. I'll be adding more details after the conference, including some benchmarking tests and unit tests.
If you've reservations around a particular set of changes or module, I can try creating a separate diff for that and discuss creating a separate PR around that.

shwstppr commented:
@blueorangutan package

blueorangutan commented:
@shwstppr a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 11549

shwstppr commented:
Will fix conflicts once we merge, https://github.com/apache/cloudstack/pull/10018

shwstppr commented on engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41910to42000.java:
```diff
@@ -58,6 +58,7 @@ public InputStream[] getPrepareScripts() {
     @Override
```
Done in b40dd4a8a418a810b5b0e909edb14bead680ec6d

shwstppr commented:
@blueorangutan package

blueorangutan commented:
@shwstppr a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✖️ el8 ✖️ el9 ✖️ debian ✖️ suse15. SL-JID 11703

shwstppr commented:
@blueorangutan package

blueorangutan commented:
@shwstppr a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✖️ el8 ✖️ el9 ✔️ debian ✖️ suse15. SL-JID 11708

BryanMLima commented on utils/src/main/java/com/cloud/utils/nio/NioClient.java:
```diff
@@ -33,73 +33,84 @@
         } catch (final IOException e) {
-            _selector.close();
+            closeChannel();
+            logger.error(String.format("IOException while connecting to %s", hostLog), e);
```
```suggestion
            logger.error("IOException while connecting to {}", hostLog, e);
```

BryanMLima commented on utils/src/main/java/com/cloud/utils/nio/NioConnection.java:
```diff
@@ -95,8 +113,8 @@ public void start() throws NioConnectionException {
         } catch (final ConnectException e) {
-            logger.warn("Unable to connect to remote: is there a server running on port " + _port);
-            return;
+            logger.warn("Unable to connect to remote: is there a server running on port" + _port, e);
```
```suggestion
            logger.warn("Unable to connect to remote: is there a server running on port {}?", _port, e);
```
Maybe even change this message to error.

BryanMLima commented on utils/src/main/java/com/cloud/utils/nio/NioConnection.java:
```diff
@@ -190,9 +209,25 @@ public Boolean call() throws NioConnectionException {
+            return false;
+        }
+        // Reject new connection if the server is busy
+        logger.warn(String.format("%s Rejecting new connection. %d active connections currently",
+                SERVER_BUSY_MESSAGE, sslHandshakeMaxWorkers));
```
```suggestion
        logger.warn("{} Rejecting new connection. {} active connections currently", SERVER_BUSY_MESSAGE, sslHandshakeMaxWorkers);
```

BryanMLima commented on utils/src/main/java/com/cloud/utils/nio/NioConnection.java:
```diff
@@ -202,57 +237,64 @@ protected void accept(final SelectionKey key) throws IOException {
+                    if (!Link.doHandshake(socketChannel, sslEngine, getSslHandshakeTimeout())) {
+                        throw new IOException("SSL handshake timed out with " + socketAddress);
+                    }
+                    if (logger.isTraceEnabled()) {
+                        logger.trace("SSL: Handshake done");
                     }
```
No need for a guard here, but if you really want to, you could use lazy logging.
```suggestion
                    logger.trace("SSL: Handshake done");
```

BryanMLima commented on utils/src/main/java/com/cloud/utils/nio/NioConnection.java:
```diff
@@ -202,57 +237,64 @@ protected void accept(final SelectionKey key) throws IOException {
+                    registerLink(socketAddress, link);
+                    _executor.submit(task);
+                } catch (final GeneralSecurityException | IOException e) {
+                    if (logger.isTraceEnabled()) {
+                        logger.trace(socket.getRemoteSocketAddress()+ "Connection closed due to failure: " + e.getMessage());
+                    }
```
```suggestion
                    logger.trace("{} Connection closed due to failure: {}", socket::getRemoteSocketAddress, e::getMessage);
```

BryanMLima commented on utils/src/main/java/com/cloud/utils/nio/NioConnection.java:
```diff
@@ -202,57 +237,64 @@ protected void accept(final SelectionKey key) throws IOException {
+                    closeAutoCloseable(socketChannel, "accepting socketChannel");
+                } finally {
+                    int connections = activeAcceptConnections.decrementAndGet();
+                    if (logger.isTraceEnabled()) {
+                        logger.trace(String.format("Accept task complete for %s - time taken: %d, " +
+                                        "active accept connections: %d",
+                                socketAddress, (System.currentTimeMillis() - startTime), connections));
+                    }
```
```suggestion
                    logger.trace("Accept task complete for {} - time taken: {}, active accept connections: {}", () -> socketAddress,
                            () -> (System.currentTimeMillis() - startTime), () -> connections);
```

BryanMLima commented on utils/src/main/java/com/cloud/utils/nio/NioConnection.java:
```diff
@@ -202,57 +237,64 @@ protected void accept(final SelectionKey key) throws IOException {
         final Link link = (Link)key.attachment();
         closeConnection(key);
         if (link != null) {
+            if (logger.isTraceEnabled()) {
+                logger.warn("Will terminate connection due to: " + msg);
+            }
```
Same here, no need for a guard
```suggestion
            logger.warn("Will terminate connection due to: {}", msg);
```

shwstppr commented on engine/schema/src/main/java/com/cloud/upgrade/dao/DatabaseAccessObject.java:
```diff
@@ -114,6 +114,17 @@ public void createIndex(Connection conn, String tableName, String indexName, Str
     }
 
+    public void renameIndex(Connection conn, String tableName, String oldName, String newName) {
+        String stmt = String.format("ALTER TABLE %s RENAME INDEX %s TO %s", tableName, oldName, newName);
```
@BryanMLima didn't get you? Are you suggesting renaming indexes from SQL upgrade paths?
Caller of this already checks if the index exists or not.

shwstppr commented on server/src/main/java/com/cloud/alert/AlertManagerImpl.java:
```diff
@@ -257,6 +259,64 @@ public void sendAlert(AlertType alertType, long dataCenterId, Long podId, String
+    }
+
+    protected void recalculateStorageCapacities() {
+        List<Long> storagePoolIds = _storagePoolDao.listAllIds();
```
@weizhouapache I may have missed this comment earlier. No, this doesn't consider states
@BryanMLima no, it won't consider removed pools.

Earlier were calling,
```
List<StoragePoolVO> storagePools = _storagePoolDao.listAll();
```
The intention here was just to get the IDs instead of all fields.

shwstppr commented:
Thanks @BryanMLima for the review. I've addressed your suggestions and responded to two queries.
In terms of testing, yes we are already testing this exhaustively with @vladimirpetrov to prevent any regressions.

@blueorangutan package

blueorangutan commented:
@shwstppr a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 11887

blueorangutan commented:
@shwstppr a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 11920

shwstppr commented:
@blueorangutan test matrix

blueorangutan commented:
@shwstppr a [SL] Trillian-Jenkins matrix job (EL8 mgmt + EL8 KVM, Ubuntu22 mgmt + Ubuntu22 KVM, EL8 mgmt + VMware 7.0u3, EL9 mgmt + XCP-ng 8.2 ) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian Build Failed (tid-11992)<b/>

blueorangutan commented:
<b>[SF] Trillian test result (tid-11993)</b>
Environment: vmware-70u3 (x2), Advanced Networking with Mgmt server ol8
Total time taken: 62635 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr9840-t11993-vmware-70u3.zip
Smoke tests completed. 137 look OK, 4 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
ContextSuite context=TestClusterDRS>:setup | `Error` | 0.00 | test_cluster_drs.py
test_list_system_vms_metrics_history | `Failure` | 0.23 | test_metrics_api.py
test_01_deployVMInSharedNetwork | `Error` | 312.96 | test_network.py
test_02_restore_vm_with_disk_offering | `Error` | 60.10 | test_restore_vm.py
test_03_restore_vm_with_disk_offering_custom_size | `Error` | 55.03 | test_restore_vm.py

blueorangutan commented:
<b>[SF] Trillian test result (tid-11991)</b>
Environment: kvm-ol8 (x2), Advanced Networking with Mgmt server ol8
Total time taken: 63469 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr9840-t11991-kvm-ol8.zip
Smoke tests completed. 140 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_secure_vm_migration | `Error` | 390.16 | test_vm_life_cycle.py
test_02_unsecure_vm_migration | `Error` | 3800.68 | test_vm_life_cycle.py
test_03_secured_to_nonsecured_vm_migration | `Error` | 378.31 | test_vm_life_cycle.py
test_04_nonsecured_to_secured_vm_migration | `Error` | 453.66 | test_vm_life_cycle.py

blueorangutan commented:
<b>[SF] Trillian test result (tid-11994)</b>
Environment: xcpng82 (x2), Advanced Networking with Mgmt server ol9
Total time taken: 74049 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr9840-t11994-xcpng82.zip
Smoke tests completed. 136 look OK, 5 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_condensed_drs_algorithm | `Failure` | 166.25 | test_cluster_drs.py
test_02_balanced_drs_algorithm | `Failure` | 183.41 | test_cluster_drs.py
ContextSuite context=TestSharedNetworkWithConfigDrive>:setup | `Error` | 10.52 | test_network.py
test_01_non_strict_host_anti_affinity | `Error` | 223.82 | test_nonstrict_affinity_group.py
test_02_non_strict_host_affinity | `Error` | 123.18 | test_nonstrict_affinity_group.py
test_02_create_volume | `Error` | 4.24 | test_resource_names.py
test_05_scale_vm_dont_allow_disk_offering_change | `Failure` | 71.36 | test_scale_vm.py

shwstppr commented:
@blueorangutan package

blueorangutan commented:
@shwstppr a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 12230

blueorangutan commented:
@rohityadavcloud a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

rohityadavcloud commented:
@blueorangutan test matrix

blueorangutan commented:
@rohityadavcloud a [SL] Trillian-Jenkins matrix job (EL8 mgmt + EL8 KVM, Ubuntu22 mgmt + Ubuntu22 KVM, EL8 mgmt + VMware 7.0u3, EL9 mgmt + XCP-ng 8.2 ) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-12253)</b>
Environment: kvm-ol8 (x2), Advanced Networking with Mgmt server ol8
Total time taken: 51652 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr9840-t12253-kvm-ol8.zip
Smoke tests completed. 140 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_11_isolated_network_with_dynamic_routed_mode | `Error` | 2.30 | test_ipv4_routing.py
test_12_vpc_and_tier_with_dynamic_routed_mode | `Error` | 2.43 | test_ipv4_routing.py
test_12_vpc_and_tier_with_dynamic_routed_mode | `Error` | 2.43 | test_ipv4_routing.py

blueorangutan commented:
<b>[SF] Trillian test result (tid-12254)</b>
Environment: kvm-ubuntu22 (x2), Advanced Networking with Mgmt server u22
Total time taken: 55866 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr9840-t12254-kvm-ubuntu22.zip
Smoke tests completed. 139 look OK, 2 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_11_isolated_network_with_dynamic_routed_mode | `Error` | 2.41 | test_ipv4_routing.py
test_12_vpc_and_tier_with_dynamic_routed_mode | `Error` | 2.52 | test_ipv4_routing.py
test_12_vpc_and_tier_with_dynamic_routed_mode | `Error` | 2.52 | test_ipv4_routing.py
test_oobm_multiple_mgmt_server_ownership | `Failure` | 30.75 | test_outofbandmanagement.py

blueorangutan commented:
<b>[SF] Trillian test result (tid-12256)</b>
Environment: xcpng82 (x2), Advanced Networking with Mgmt server ol9
Total time taken: 79994 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr9840-t12256-xcpng82.zip
Smoke tests completed. 135 look OK, 6 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_condensed_drs_algorithm | `Failure` | 168.36 | test_cluster_drs.py
test_02_balanced_drs_algorithm | `Failure` | 185.49 | test_cluster_drs.py
test_11_isolated_network_with_dynamic_routed_mode | `Error` | 1.30 | test_ipv4_routing.py
test_12_vpc_and_tier_with_dynamic_routed_mode | `Error` | 2.46 | test_ipv4_routing.py
test_12_vpc_and_tier_with_dynamic_routed_mode | `Error` | 2.47 | test_ipv4_routing.py
ContextSuite context=TestSharedNetworkWithConfigDrive>:setup | `Error` | 7.42 | test_network.py
test_01_non_strict_host_anti_affinity | `Error` | 217.06 | test_nonstrict_affinity_group.py
test_02_non_strict_host_affinity | `Error` | 113.29 | test_nonstrict_affinity_group.py
test_02_create_volume | `Error` | 4.28 | test_resource_names.py
test_05_scale_vm_dont_allow_disk_offering_change | `Failure` | 65.77 | test_scale_vm.py
